### PR TITLE
add csv fixture for e2e export/import test

### PR DESCRIPTION
**Pull request by ruslandoga**
### Changes

This PR adds CSVs with some data from plausible.io/plausible.io and starts using them in e2e export/import test.

- [x] don't inline csv but put them into some files
- [x] perform comparisons between native and imported queries

### Tests
- [x] Automated tests have been updated

### Changelog
- [x] This PR does not make a user-facing change

### Documentation
- [x] This change does not need a documentation update

### Dark mode
- [x] This PR does not change the UI

### CODE REVIEW

File: test/plausible/imported/csv_importer_test.exs [hunk 1]
```diff
@@ -412,78 +413,44 @@ defmodule Plausible.Imported.CSVImporterTest do
...
-          operating_system: "MacNoVersion"
-        ),
-        build(:event,
-          timestamp:
-            Timex.shift(~N[2021-10-20 12:00:00], days: -1) |> NaiveDateTime.truncate(:second),
-          name: "Signup",
-          "meta.key": ["variant"],
-          "meta.value": ["A"]
-        )
+    test "it works", %{conn: conn, user: user, tmp_dir: tmp_dir} do
```
Comment (ruslandoga) [inline test/plausible/imported/csv_importer_test.exs:419]: it works with #4035 and #4051

File: test/plausible/imported/csv_importer_test.exs [hunk 2]
```diff
@@ -533,13 +502,564 @@ defmodule Plausible.Imported.CSVImporterTest do
...
+
+      exported_timeseries = timeseries.(timeseries_params.(exported_site))
+      imported_timeseries = timeseries.(timeseries_params.(imported_site))
+
+      pairwise(exported_timeseries, imported_timeseries, fn exported, imported ->
+        assert exported["date"] == imported["date"]
+        assert exported["pageviews"] == imported["pageviews"]
+      end)
+
+      # NOTE: timeseries' views per visit difference is up to 160%
```
Comment (RobertJoonas) [inline test/plausible/imported/csv_importer_test.exs:567]: That's a huge difference, but we probably shouldn't think of the native `views_per_visit` as the source of truth either. How the imported `views_per_visit` is calculated actually seems more reliable.

The native query is calculating it as `s.sign * s.pageviews) / sum(s.sign)`. which I think is more likely to have something fishy going on. I guess that will be handled separately - not a blocker for CSV imports/exports.

File: test/plausible/imported/csv_importer_test.exs [hunk 3]
```diff
@@ -533,13 +502,438 @@ defmodule Plausible.Imported.CSVImporterTest do
...
+
+      # city breakdown's bounce_rate difference is within 5
+      assert summary(field(exported_cities, "bounce_rate")) == [0, 17, 50, 100, 100]
+      assert summary(field(imported_cities, "bounce_rate")) == [0, 17, 50, 100, 100]
+
+      pairwise(exported_cities, imported_cities, fn exported, imported ->
+        e = exported["bounce_rate"]
+        i = imported["bounce_rate"]
+
+        if is_number(e) and is_number(i) do
```
Comment (RobertJoonas) [inline test/plausible/imported/csv_importer_test.exs:810]: In what situation is bounce rate not a number? I've tried running the test locally, and asserting on bounce rate equality between imported and exported:

`assert exported["bounce_rate"] == imported["bounce_rate"]`

The test is passing with `experimental_reduced_joins` on and off. Should we just assert equality then?

**Review by RobertJoonas — APPROVED** [review]

File: test/plausible/imported/csv_importer_test.exs (hunk 3 above)
Comment (ruslandoga) [reply test/plausible/imported/csv_importer_test.exs:810]: Yes, I think this "pairwise" block was a mistake. Bounce rate is asserted equal in all other breakdowns.

Comment (ruslandoga) [reply test/plausible/imported/csv_importer_test.exs]: https://github.com/plausible/analytics/pull/4037/commits/6a10ad994eef50a73eb74ff4e0abbbe16bd8ce90

File: test/plausible/imported/csv_importer_test.exs (hunk 2 above)
Comment (ruslandoga) [reply test/plausible/imported/csv_importer_test.exs:578]: The formula seems correct. I think the problem is with the data and how sessions are collapsed.

```sql
select sum(sign * pageviews) from sessions_v2;
┌─sum(multiply(sign, pageviews))─┐
│                           6325 │
└────────────────────────────────┘
select count() from events_v2 where name = 'pageview';
┌─count()─┐
│   12250 │
└─────────┘
select s.session_id, e.session_id, s.pageviews, e.pageviews
from (select session_id, sum(pageviews) as pageviews from sessions_v2 group by session_id) s
full join (select session_id, count() as pageviews from events_v2 where name = 'pageview' group by session_id) e 
on s.session_id = e.session_id
having s.pageviews != e.pageviews;
┌───────────session_id─┬─────────e.session_id─┬─pageviews─┬─e.pageviews─┐
│   274039145453238484 │   274039145453238484 │         2 │           1 │
│  1966893030056597902 │  1966893030056597902 │        29 │          27 │
│ 14367321820128290977 │ 14367321820128290977 │         5 │           3 │
└──────────────────────┴──────────────────────┴───────────┴─────────────┘
┌─session_id─┬─────────e.session_id─┬─pageviews─┬─e.pageviews─┐
│          0 │   909729459580268341 │         0 │           1 │
│          0 │  5041567837619101952 │         0 │           5 │
│          0 │  9941822201751237677 │         0 │         269 │
│          0 │ 11179857592511640054 │         0 │           6 │
│          0 │  7914293758317123167 │         0 │         349 │
│          0 │  2843014593483252784 │         0 │           2 │
│          0 │ 16120773483696094616 │         0 │          24 │
│          0 │  1082359377393852532 │         0 │           5 │
│          0 │  9083812924395524907 │         0 │           3 │
│          0 │   564929186898220874 │         0 │        5252 │
│          0 │ 15890328459878810482 │         0 │          10 │
└────────────┴──────────────────────┴───────────┴─────────────┘
```

Also there is someone who got cancelled before being counted:

```sql
select session_id, groupArray(sign) as signs from sessions_v2 s group by session_id having sum(sign) < 1;
┌──────────session_id─┬─signs──┐
│ 1966893030056597902 │ [-1,1] │
└─────────────────────┴────────┘
select * from sessions_v2 where session_id = 1966893030056597902 format Vertical;
Row 1:
──────
session_id:               1966893030056597902
sign:                     -1
site_id:                  236386
user_id:                  2024348104656530947
hostname:                 plausible.io
timestamp:                2024-03-31 20:29:25
start:                    2024-03-31 20:29:25
is_bounce:                1
entry_page:               /sites
exit_page:                /sites
pageviews:                1
events:                   1
duration:                 0
referrer:
referrer_source:
country_code:             FR
screen_size:              Desktop
operating_system:         Mac
browser:                  Chrome
utm_medium:
utm_source:
utm_campaign:
browser_version:          123.0
operating_system_version: 10.15
subdivision1_code:        FR-PDL
subdivision2_code:        FR-44
city_geoname_id:          2990969
utm_content:
utm_term:
transferred_from:
entry_meta.key:           ['browser_language','logged_in','theme']
entry_meta.value:         ['fr','true','system']
exit_page_hostname:

Row 2:
──────
session_id:               1966893030056597902
sign:                     1
site_id:                  236386
user_id:                  2024348104656530947
hostname:                 plausible.io
timestamp:                2024-03-31 22:42:20
start:                    2024-03-31 20:29:25
is_bounce:                0
entry_page:               /sites
exit_page:                /:dashboard
pageviews:                28
events:                   28
duration:                 7975
referrer:
referrer_source:
country_code:             FR
screen_size:              Desktop
operating_system:         Mac
browser:                  Chrome
utm_medium:
utm_source:
utm_campaign:
browser_version:          123.0
operating_system_version: 10.15
subdivision1_code:        FR-PDL
subdivision2_code:        FR-44
city_geoname_id:          2990969
utm_content:
utm_term:
transferred_from:
entry_meta.key:           ['browser_language','logged_in','theme']
entry_meta.value:         ['fr','true','system']
exit_page_hostname:
```
For that reason they are not included in the denominator

```sql
select sum(sign) from sessions_v2;
┌─sum(sign)─┐
│      2087 │
└───────────┘
select uniq(session_id) from sessions_v2;
┌─uniq(session_id)─┐
│             2088 │
└──────────────────┘
```

Comment (ruslandoga) [reply test/plausible/imported/csv_importer_test.exs:578]: Note that 1966893030056597902 is interesting as it shows up in both. Seems like that user had 29 pageviews but the session was collapsed wrong, it includes the updates after their first and 28th pageview.
```sql
select sign, pageviews from sessions_v2 where session_id = 1966893030056597902;
-- ┌─sign─┬─pageviews─┐
-- │   -1 │         1 │
-- │    1 │        28 │
-- └──────┴───────────┘
```

Comment (ruslandoga) [reply test/plausible/imported/csv_importer_test.exs:578]: 274039145453238484 is also interesting. It counts sessions_v2.pageviews for a session_id that 's not used in that user's events_v2.session_id.

```sql
select sign, user_id, pageviews, start, timestamp from sessions_v2 where session_id = 274039145453238484;
-- ┌─sign─┬─────────────user_id─┬─pageviews─┬───────────────start─┬───────────timestamp─┐
-- │    1 │ 3304869223228810334 │         2 │ 2024-03-30 23:59:46 │ 2024-03-31 00:25:24 │
-- └──────┴─────────────────────┴───────────┴─────────────────────┴─────────────────────┘
select timestamp, name, user_id from events_v2 where session_id = 274039145453238484;
-- ┌───────────timestamp─┬─name─────┬─────────────user_id─┐
-- │ 2024-03-31 00:25:24 │ pageview │ 3304869223228810334 │
-- └─────────────────────┴──────────┴─────────────────────┘
select timestamp, name, session_id from events_v2 where user_id = 3304869223228810334;
-- ┌───────────timestamp─┬─name─────┬───────────session_id─┐
-- │ 2024-03-31 00:25:24 │ pageview │   274039145453238484 │
-- │ 2024-03-31 08:51:40 │ pageview │    21278701133733325 │
-- │ 2024-03-31 08:51:42 │ pageview │    21278701133733325 │
-- │ 2024-03-31 08:52:33 │ pageview │    21278701133733325 │
-- │ 2024-03-31 08:52:37 │ pageview │    21278701133733325 │
-- │ 2024-03-31 13:27:53 │ pageview │  7143492776072111708 │
-- │ 2024-03-31 14:29:54 │ pageview │  5798352593132217646 │
-- │ 2024-03-31 15:57:21 │ pageview │  9156717935310957786 │
-- │ 2024-03-31 17:23:41 │ pageview │  7847180938336737359 │
-- │ 2024-03-31 17:31:23 │ pageview │  7847180938336737359 │
-- │ 2024-03-31 18:11:25 │ pageview │    90220698885170028 │
-- │ 2024-03-31 19:23:22 │ pageview │  5296168014377302302 │
-- │ 2024-03-31 20:45:31 │ pageview │ 17458776404432328553 │
-- │ 2024-03-31 20:45:33 │ pageview │ 17458776404432328553 │
-- │ 2024-03-31 21:24:07 │ pageview │  3820428654961765653 │
-- │ 2024-03-31 22:16:08 │ pageview │ 11967332617232776699 │
-- │ 2024-03-31 22:46:38 │ pageview │ 18026232568293063616 │
-- │ 2024-03-31 23:28:21 │ pageview │  4812138914817681207 │
-- │ 2024-03-31 23:29:18 │ pageview │  4812138914817681207 │
-- └─────────────────────┴──────────┴──────────────────────┘
```

Comment (RobertJoonas) [reply test/plausible/imported/csv_importer_test.exs:578]: Thanks for the investigation! Will definitely be useful to whoever will pick it up 🙂

Comment (ruslandoga) [reply test/plausible/imported/csv_importer_test.exs:578]: I think this also means that salts rotation might be working but session extension and user id update in session makes it seem that it doesn't.

**Pull request merged**